docs/cip/CIP-8_Matrix_Packages.md

```diff
@@ -0,0 +1,209 @@
+
+##### `data.embedding`
+```yaml
+  embedding:
+    model: "hf://sentence-transformers/all-MiniLM-L6-v2"
+    types:
+      dimensions: 384
+      quantize:
+        bits: 8
+        method: "k-means"
+```
+
+The purpose of this stanza is to describe type-like properties of the vector space for the embeddings in the package. It is an attempt at "where, not how."
+
+`embedding.model` tells us the model that produced the embeddings in the data file. A simple string identifier for the model that produced the embeddings is sufficient here to establish what amounts to a [nominal type system](https://medium.com/@thejameskyle/type-systems-structural-vs-nominal-typing-explained-56511dd969f4). The name of the model here can then be thought of as the name of the universe of weights defined by the trained model, and by extension the computed embeddings.
+
+There is plenty of debate on the virtues of structural vs. nominal typing in the traditional software world. It often amounts to, "is it better DX to assume that 'if it quacks like a duck, it's a `Duck`' (structural), or to require explicit opt-in on a definition of `Duck`? (nominal)" Nominal type systems also implicitly allow names to carry opaque, special meaning beyond their formally expressed properties. That opacity makes nominal typing the natural choice for this case, as it mirrors the opacity of a trained model's weights.
+
+`embedding` is closely related to, but more constrained than Chroma's embedding function - it names only the model, rather some known invokable function whose (opaque) behavior is to call that model.
+
+TODO is this over-abstracting the notion of an embedding func, to the point where it's not useful at all?
+
+`embedding.types` describes other knowable properties about the vector space in which the embeddings live - what's the dimensionality of the space? is it quantized? For Chroma's purposes, these descriptive properties should be treated as additional constraints on the [query pipeline](https://github.com/chroma-core/chroma/blob/6384b66bc19e6a6217c2c5aa55f3b539e30c2708/docs/CIP_6_Pipelines_Registry.md#query-pipelines) that must be applied when processing an input for vector search.
+
+TODO i am totally out of my depth with quantization and need to run this through with someone. Key questions - does it make sense to try to reduce embedding and query pipeline fns to a model + some set of constraint arguments? Is quantization distinct from embedding - i.e. do we have one column for each in the parquet file?
+TODO finish
+
+#### `data.digests`
+```yaml
+  digests:
+    data: "sha256:decafbad"
+```
+
+### `matrix.data.parquet`
+
+The `matrix.data.parquet` file contains the core dataset of the matrix package. Each row in this Parquet file closely mirrors Chroma database content. The following columns are expected:
```

Comment:
to retain fidelity with the source system we can also include `id` as of right now this is user-defined.

Reply:
Yeah, this is interesting.

I had included `id` in the first pass at this writeup. I removed it because of...a hard-to-articulate intuition, hoping it would provoke discussion that we could use to nail it down.

To that end, and assuming that Chroma's API/client treats ids purely as a unique identifier for documents without any additional semantics: say we include `id` in matrix packages, and Ahmaud publishes a matrix package containing 100 documents. Breonna pulls this package into a collection in her own Chroma, which contains exactly one `id` from the package. What should the behavior of the importer be?

(yes, this goes beyond the simplifying "1:1 relation between matrix package and collection" premise of this initial pass)

Reply:
i percolated on this a fair bit yesterday. i think we can actually talk about it quite concretely in the context of `matrix-import` and `matrix-export`, so i'm working on updates to the doc in those sections.

i do suspect, however, that a proper solution here may require some (simple) schema changes to Chroma in order to track the mapping between data held in Chroma and some external matrix package. The basic purpose of such field(s) would be:

1. Clearly differentiating between Chroma db contents that are from a matrix package (comparable to `/usr/lib` or `node_modules` in trad pkg mgmt) vs local and modifiable (comparable to your project's code).
2. Facilitating export of a matrix package from db contents

i am hoping that actually hacking at the CLI implementation will yield a minimal approach, good enough for this first pass, that takes advantage of the 1:1 collection to matrix package simplification.

Reply:
We can leave the choice in the user's hands by giving them --ignore-ids on export or --regenerate-ids on import. This way we can differ this design choice (aka kick the can down the road).